### DISCUSSION TITLE
GPS over LXMF (request for conceptual assistance)

### DISCUSSION THREAD

**Discussion by faragher** [markqvist/Reticulum#232] [Help & Questions] [answered]
One of the things I've been excited about is the ability to use personal equipment to assist search and rescue people in the field. Obviously, there's an outpouring of well-meaning but often undertrained and underequipped volunteers when there are people lost in the wilderness. By using a tracking system at incident command, the chances of areas being missed or searchers getting lost is minimized. 

LoRA uplinks are far superior to WiFi and often superior to cell signals in many of these regions, particularly in these mesh configurations or if an airborne platform can house a node.

I've had excellent luck with Traccar, which accepts a wide variety of GPS formats as input over TCP/IP but I don't want to rely on running IP over Reticulum when direct communication is possible. I'd like to float the general process I was thinking of, mostly in the hopes of not running around in circles duplicating effort or making an inappropriate system. 

**Private Phone:** Using an LXMF compatible client, GPS data is pulled from the phone, and sent in a compatible GPS format (with or without additional headers) to a pre-shared address using an RNode connected via bluetooth.

**RNode:** Does the magic of sending the data over LoRA. I was considering a custom hardware and modified firmware to take care of GPS positioning on the RNode itself, but given the near ubiquitous GPS on cell phones, and the fact that a $20 cell phone could be pressed into service for close to the cost of the additional board space and module, I'm not sure it's worth it. Especially forking the hardware and firmware. 

**Server:** Python LXMF client sees the packet, (from sever RNode or other hardware interface) ensures it's meant for the node, authenticates the address as a registered node, and then sends a TCP/IP request to Traccar, entering the data in that server. 

It's basically sending an arbitrary data packet, but I'd like to make sure I'm not missing something big or misunderstanding the process. I'm not a mobile dev, so I would prefer not to try to troubleshoot both the concept and learning the peculiarities of mobile at the same time. I'd also very much like to not duplicate effort if someone else has done similar work.

Thanks in advance!

**Comment by markqvist** [comment, marked as answer]
Hi @faragher!

This is a really valuable area of application, and that I've hoped for a long time Reticulum would find it's use in, since it is so well suited to such scenarios.

The overall architecture of the approach you outline is sound. But this is a case where I would actually say that using LXMF is too much overhead, unless you need to make sure that *all* position reports are ultimately received by the coordination center.

If this is not the case, a better approach would be to just use the raw `Packet` API. Since all such position or telemetry reports would easily fit into the packet MTU of Reticulum, they can just be efficiently sent as singular packets over the network to their final destination. This model is also very suitable for situations where you have a large amount of people moving around, possibly dropping in and out of coverage, and transmitting telemetry in an uncoordinated fashion. You don't even waste bandwidth on establishing and maintaining Reticulum Links.

Such a system easily scales to a lot of people, even on low bandwidth mediums. You can also segment devices out into different channels to increase capacity further.

I wish there was already an Android basic app source example available. In that case, it would be quite easy to just add some simple code to send the telemetry every X seconds, or once the position has changed more than a defined threshold. Unfortunately, no such "quickstart project" exists yet. You may be able to use the Sideband app as a starting point though.

Another option would be to just add the functionality to Sideband. It was the plan from the beginning to add this kind of functionality, but it has not been completed yet.

I hope this all gives a little more clarity to the subject!

**Reply by faragher** [reply]
To clarify, the packet API doesn't require a Reticulum instance? That would simplify the ecosystem.

But if there's an LXMF implementation, would it compete with the Reticulum stack in Sideband? 

The more I think about it the more I suggest that a beacon functionality be added to Sideband eventually, and perhaps a Packet based beacon be developed as a lighter alternative. Since it seems simpler as a starting place, I think that's the idea I'll pursue. If nothing else, it would be easy enough to re-use the code if its developed from the same libraries on the same hardware.

Thanks!

**Reply by markqvist** [reply]
> But if there's an LXMF implementation, would it compete with the Reticulum stack in Sideband?

I don't understand this question. Can you clarify or elaborate on what you mean?

> To clarify, the packet API doesn't require a Reticulum instance? That would simplify the ecosystem.

It does require a full transport-enabled Reticulum instance to *transport* over multiple hops in the network. But since the Packet API is very simple, it will be one of the first things to be added to a microcontroller-compatible Reticulum C library. In the relative near future, it will be possible to at least use the Packet APIs from simple microcontroller-based devices. Full instances (for example running on a computer, Raspberry Pi or Android device) can then transport these packets over any number of hops in the network.

The simplest path to a working system *right now* would be to add GPS position beaconing to Sideband, by using the RNS Packet API (and not LXMF; it is more efficient to just use raw packets for this). Then write a simple program to receive the position reports in the other end and export them to GPX tracks or similar (less than 100 lines of code probably).

**Reply by faragher** [reply]
> > But if there's an LXMF implementation, would it compete with the Reticulum stack in Sideband?
> 
> I don't understand this question. Can you clarify or elaborate on what you mean?

Ignorance of the network stack on my part. I was unclear on the fact that while each program has an individual Reticulum instance, they all communicate and can even use a daemon to avoid drop in/out affecting the other systems. While I might not be 100% correct, it seems this would have never been a problem.


> It does require a full transport-enabled Reticulum instance to _transport_ over multiple hops in the network. But since the Packet API is very simple, it will be one of the first things to be added to a microcontroller-compatible Reticulum C library. In the relative near future, it will be possible to at least use the Packet APIs from simple microcontroller-based devices. Full instances (for example running on a computer, Raspberry Pi or Android device) can then transport these packets over any number of hops in the network.

Got it, a well formed packet is able to go point to point, and (soon) a full implementation could route it, making "dumb" broadcast units capable of routing through other nodes.
 
> The simplest path to a working system _right now_ would be to add GPS position beaconing to Sideband, by using the RNS Packet API (and not LXMF; it is more efficient to just use raw packets for this). Then write a simple program to receive the position reports in the other end and export them to GPX tracks or similar (less than 100 lines of code probably).

Got it. I have a testbed I've been working on since yesterday. It's currently loading/saving configurations and broadcasting. It's provided me a much better understanding of the stack. When I get it to MVP I'll take a step back and see if I can make it work in Sideband. I was also thinking of using QR codes to ingest configuration data, which could be considered dangerous. Is that something you'd like to avoid in Sideband?

It's a pleasure working with this API, by the way. The only real question I have is: it appears I can change the radio configuration on the fly without restart. Is there any reason I shouldn't? Does it only appear to work because I was changing broadcast power?

**Reply by markqvist** [reply]
> I was also thinking of using QR codes to ingest configuration data, which could be considered dangerous. Is that something you'd like to avoid in Sideband?

Would be OK, as long as the user is able to verify the ingested parameters before actually comitting them to their active configuration.

> It's a pleasure working with this API, by the way.

Thanks, that is awesome to hear, since that was the design intent :)

> The only real question I have is: it appears I can change the radio configuration on the fly without restart. Is there any reason I shouldn't? Does it only appear to work because I was changing broadcast power?

You should be able to do this without issue. It is not something that was meant as anything you would *need* to do in most cases, but I left the API open to doing stuff like that, in the case that anyone would find it useful :)

**Reply by buhalo1** [reply]
Could you add the ability to work with APRS to the reticulum? This is a fairly advanced protocol and there are a significant number of popular applications working with this protocol.
The ideal option would be to implement the ability to connect aprs applications to reticulum via kissip modem emulation, with further routing of APRS packets within reticulum interfaces or reticulum network in accordance with the route headers included to aprs packet transmitted by the application.

The following route headers could be used:

Common:
1) BEACON - APRS BEACON
2) WIDEN-N DIGIPEATERS
3) GATEnn - CROSSBAND/CROSSMODE
4) HFnn - HF GATEWAY

Reticulum:
5) BCAST - RETICULUM BROADCAST
6) MCAST - RETICULUM MULTICAST TO APRS GROUP
7) IGATE - RETICULUM TO APRS-IS

On routes 1-2, the packet is transmitted by the reticulum interfaces in native APRS format without reticulum overhead.

A packet transmitted without a route header is not relayed by the hosts that received it, but is routed to the APRS-IS server via transport network channels.

A packet with a route header of 1 is not relayed by the hosts that received it, but is routed to the APRS-IS server over the transport network channels or via directed transmission over the reticulum network.

Retransmission of a packet with route header 2 and processing of the route header by digipeater nodes is carried out according to the routing rules in APRS networks.

Routing of a packet with route header 3 by nodes that do not support communication over the reticulum network is carried out in the native APRS format directly to the HF gateway that supports operation in band and mode specified by the digital suffix of the route header, or to the nearest node that supports communication over the reticulum network and has a connection in this network with a HF gateway that supports operation on a band and mode specified by the digital suffix of the route header, after which the packet is directed to this HF gateway over the reticulum network.
The first digit in the numeric suffix of the route header indicates the HF band, the second is the packet transmission mode (AFSK, PSK, etc.).

Routing a packet with a route header of 4 nodes that do not support communication over the reticulum network is carried out in the native APRS format directly to the nodes that provide packet retransmission to APRS subnet operating in other bands and modes or to the nearest node that supports communication over the reticulum network.
Upon receipt of a packet with the given route header, the node that supports communication over the reticulum network performs its directed transmission to the nearest gateways that provide its retransmission to the APRS subnet operating in other bands and modes.
The first digit in the numeric suffix of the route header indicates the VHF band and the second is the mode in which the packet was originally transmitted.

On routes 5-7, the packet is transmitted over the reticulum network.
The packet with route header 5 is transmitted in BROADCAST mode.
The packet with route header 6 is transmitted in multicast mode for the APRS group.
The packet with route header 7 is transmitted in directed mode to the nearest APRS-IS gateway.

**Reply by faragher** [reply]
Is this something that requires changes to Reticulum? The beacon I'm working on works just fine on the stack without modification and data is data. Given its ability to hook into packet radios (right?) it might just be an issue of writing software to manage the messages and pass it to whatever handles the APRS input/output.

**Reply by buhalo1** [reply]
I am currently trying to build a system with the described functionality. This requires solving several problems.
1) KISS TCP modem emulator to which amateur radio applications would connect.

For these purposes, you can use TCP KISS server https://github.com/simplyequipped/tcpkissserver

2) A parser-router that would decode packets from amateur applications arriving at the TCP KISS server would determine and process the PATH headers contained in these packets, after which, depending on the header, it would send these packets along two different routes.

a) Through the reticulum in broadcast, multicast or directional modes.

b) Bypassing the reticulum, but to the same interfaces used for the reticulum. In this case, I'm primarily interested in routing such packets to KISS (RNODE & AFSK) and UDP broadcast (DMR & P25 radios) interfaces used by reticulum, in the original format.
It would be even better to be able to choose which interface to send the packet with one or another PATH header, and in the case of transmission via UDP BROADCAST, select the port through which this packet will be sent depending on the PATH header included in the packet or on whether through which connection to the TCP PATH server the packet was received on.

Problem #1 is the parser-router. Programming is not my forte, the most I can do is to intuitively read and, if necessary, edit someone else's code a little, and also sometimes admire the harmony of someone else's code, for example, the reticulum code. But in any case, this task can be solved by a wide range of programmers, since its solution does not require specific knowledge.

A more difficult task, in my opinion, is the transfer of packets bypassing the reticulum but through the interfaces used by the reticulum. In my understanding, for these purposes, a bypass interface should be provided in the reticulum, which would pass packets arriving on it to the interfaces used by the reticulum without any processing.

**Reply by markqvist** [reply]
There is no need to pass Reticulum traffic over APRS. The APRS protocol is not very advanced, and has no native routing support. It would only add tons of overhead and make things much more inefficient.

You can already pass Reticulum traffic directly over AFSK packet modems, or any other systems using such hardware. No point in wrapping it all in APRS packets.

If you *really*, *really* need to, there is already support for the AX.25 interfaces, that allows you to pass Reticulum traffic over AX.25 networks, but I honestly don't recommend that, unless you absolutely cannot avoid it.

**Reply by buhalo1** [reply]
@markqvist I have the opposite goal. The goal is to transmit through interfaces used by reticulum, data from applications connected via tcp kiss in their native format without reticulum overhead. But still being able to use the reticulum.
I am considering two such applications.
1 is APRS, 2 is chattervox. This possibility is necessary for communication with operators who do not use reticulum. For example, chattervox allows you to broadcast a link to a website to all participants during an evening net over repeater, even those who do not use reticulum, but use, for example, a simple AFSK modem and an amateur radio application connected to it.
But also, for example, it can be a classic AX25 BBS running over UDP BROADCAST over P25 radios.
This option will allow you to use dongles based on vocore2 or rpi0w connected to the radios, while all of the listed applications can simultaneously run on computers connected to these dongles over wifi.

**Reply by faragher** [reply]
>without reticulum overhead.

Sending arbitrary data over the RNode is trivial, if that's what you're looking for. Otherwise I'm not sure exactly how you wish to use Reticulum interfaces without Reticulum itself and its associated overhead.

Fire up Python and play with RNS and you should be able to get raw data to/from a LoRa RNode, otherwise I don't understand what you're looking for.
